token/src/instruction.rs

```diff
@@ -22,69 +22,66 @@ pub struct TokenInfo {
 #[repr(C)]
 #[derive(Clone, Debug, PartialEq)]
 pub enum TokenInstruction {
-    /// Initializes a new mint and deposits all the newly minted tokens in an account.
+    /// Initializes a new mint and optionally deposits all the newly minted tokens in an account.
     ///
     /// # Accounts expected by this instruction:
     ///
     ///   0. `[writable, signer]` New mint to create.
     ///   1.
     ///      * If supply is non-zero: `[writable]` Account to hold all the newly minted tokens.
     ///      * If supply is zero: `[]` Owner of the mint.
-    ///   2. Optional: `[]` Owner of the mint if supply is non-zero, if present then the 
-    ///      token allows further minting of tokens.
+    ///   2. Optional: `[]` Owner of the mint if supply is non-zero, if present then further
+    ///      minting is supported.
     InitializeMint(TokenInfo),
-    /// Initializes a new account.  The new account can either hold tokens or be a delegate
-    /// for another account.
+    /// Initializes a new account.
     ///
     /// # Accounts expected by this instruction:
     ///
-    ///   0. `[writable, signer]`  New account being created.
-    ///   1. `[]` Owner of the new account.
-    ///   2. `[]` Token this account will be associated with.
-    ///   3. Optional: `[]` Source account that this account will be a delegate for.
+    ///   0. `[writable, signer]`  New account being initialized.
+    ///   1. `[]` The mint this account will be associated with.
+    ///   2. `[]` Owner of the new account.
     InitializeAccount,
     /// Transfers tokens from one account to another either directly or via a delegate.
     ///
     /// # Accounts expected by this instruction:
     ///
-    ///   0. `[signer]` Owner of the source account.
-    ///   1. `[writable]` Source/Delegate account.
-    ///   2. `[writable]` Destination account.
-    ///   3. Optional: `[writable]` Source account if key 1 is a delegate account.
+    ///   0. `[writable]` The source account.
+    ///   1. `[writable]` The destination account.
+    ///   2. '[signer]' The source's owner or delegate
     Transfer(u64),
-    /// Approves a delegate.  A delegate account is given the authority to transfer
-    /// another accounts tokens without the other account's owner signing the transfer.
+    /// Approves a delegate.  A delegate is given the authority over
+    /// tokens on behalf of the source account's owner.  If the amount to
+    /// delegate is zero then delegation is rescinded
     ///
     /// # Accounts expected by this instruction:
     ///
-    ///   0. `[signer]` Owner of the source account.
-    ///   1. `[]` Source account.
-    ///   2. `[writable]` Delegate account.
+    ///   0. `[writable]` The source account.
+    ///   1. Optional: `[writable]` The delegate if amount is non-zero.
+    ///   2. `[signer]`The source account owner address
```

Comment:
Just a nit:  make the optional account the 3rd account?  I think that's generally easier to code up, and also express in the documentation.    
As a motivating example, the comment "2. `[signer]`The source account owner address" is incomplete because the source account could also be at position 1 if no delegate is present.

Reply:
@mvines Yeah, approve was the weird one trying to support "Revoke" because of the optional parameter but I left it this way since I was about to add multisig which would result in a similar "motivating example".  Also, if the accounts are 1, 2, 3, and 2 is optional it's probably fair to expect folks to understand that 3 moves to 2 if optional 2 is not included.  I think the real answer here is we should avoid optional as much as possible and a separate revoke instruction makes more sense.